src/wasm/module-decoder.cc

```diff
@@ -116,10 +116,17 @@ class ModuleDecoder : public Decoder {
                 {nullptr, 0, 0, 0, 0, 0, 0, false, false});
             WasmFunction* function = &module->functions->back();
             DecodeFunctionInModule(module, function, verify_functions);
-
-            if (ok() && verify_functions) {
-              if (!function->external)
+          }
+          if (ok() && verify_functions) {
```

Comment:
Hi titzer, the separated two iterations here actually are used to allow function callings pass the verification. In the one-iteration edition, if `foo1` calls `foo2` in its function body, but `foo2` is written after `foo1`, then `VerifyFunctionBody` for `foo1` will fail, because `foo2` is not decoded and registered to `module->functions` at that time.
The `if (ok() && verify_functions)` and `if (failed()) break;` checks will terminate the verification as soon as anything fails.